pkg/lib/cockpit/location.ts

```diff
@@ -0,0 +1,192 @@
+        }
+
+        this.href = get_window_location_hash();
+        this.options = {};
+        this.path = this.decode(this.href, this.options);
+    }
+
+    #resolve_path_dots(parts: string[]) {
+        const out = [];
+        for (let i = 0; i < parts.length; i++) {
+            const part = parts[i];
+            if (part === "" || part == ".") {
+                continue;
+            } else if (part == "..") {
+                if (out.length === 0)
+                    return [];
+                out.pop();
+            } else {
+                out.push(part);
+            }
+        }
+        return out;
+    }
+
+    #href_for_go_or_replace(path: Path, options?: Options) {
+        options = options || {};
+        if (typeof path === "string") {
+            return this.encode(this.decode(path, options), options);
+        } else if (path instanceof Location) {
+            return path.href;
+        } else {
+            return this.encode(path, options);
+        }
+    }
+
+    #decode_path(input: string) {
+        let result, i;
+        let pre_parts: string[] = [];
+        const parts = input.split('/').map(decodeURIComponent);
+
+        if (this.url_root)
+            pre_parts = this.url_root.split('/').map(decodeURIComponent);
+
+        if (input && input[0] !== "/" && this.path !== undefined) {
+            result = [...this.path];
+            result.pop();
+            result = result.concat(parts);
+        } else {
+            result = parts;
+        }
+
+        result = this.#resolve_path_dots(result);
+        for (i = 0; i < pre_parts.length; i++) {
+            if (pre_parts[i] !== result[i])
+                break;
+        }
+        if (i == pre_parts.length)
+            result.splice(0, pre_parts.length);
+
+        return result;
+    }
+
+    encode(path: string | string[], options: Options, with_root: boolean = false) {
+        if (typeof path == "string")
+            path = this.#decode_path(path);
+
+        let href = "/" + path.map(encodeURIComponent).join("/");
+        if (with_root && this.url_root && href.indexOf("/" + this.url_root + "/") !== 0)
+            href = "/" + this.url_root + href;
+
+        /* Undo unnecessary encoding of these */
+        href = href.replaceAll("%40", "@");
+        href = href.replaceAll("%3D", "=");
+        href = href.replaceAll("%2B", "+");
+        href = href.replaceAll("%23", "#");
+
+        const query: string[] = [];
+        if (options) {
+            for (const opt in options) {
+                let value = options[opt];
+                if (!Array.isArray(value))
+                    value = [value];
+                value.forEach(function(v: string) {
+                    query.push(encodeURIComponent(opt) + "=" + encodeURIComponent(v));
+                });
+            }
+            if (query.length > 0)
+                href += "?" + query.join("&");
+        }
+        return href;
+    }
+
+    decode(href: string, options: Options) {
```

Comment:
As public function, the return type is especially important here. I would expect a `Location` and this to be a static top-level function, or a constructor. But that's apparently not what it is.

Reading https://cockpit-project.org/guide/latest/cockpit-location.html, I suppose it returns a `string[]`, and `options` is an *out* argument, not an input? A description comment here is appreciated. Of course it'd be better to not have out arguments in the first place, and instead return a `{ path: string[], options: Options }`, but that'd break API :cry:

Reply:
Looking at our tests, we indeed test this behaviour :exploding_head: 
```
const checks = [
        ["relative/../sub", {
            path: ["base", "sub"],
            options: { }
        }],
]
...
        const path = cockpit.location.decode(checks[i][0], options);
        assert.deepEqual({ path, options }, checks[i][1], "decode(\"" + checks[i][0] + "\")");

```